packages/components/src/input-control/input-field.tsx

```diff
@@ -105,14 +105,13 @@ function InputField(
 		onBlur( event );
 		setIsFocused?.( false );
 
-		/**
-		 * If isPressEnterToChange is set, this commits the value to
-		 * the onChange callback.
-		 */
+		// If isPressEnterToChange is set, cause the `onChange` callback to be fired.
 		if ( isPressEnterToChange && isDirty ) {
 			wasDirtyOnBlur.current = true;
-			handleOnCommit( event );
 		}
+
+		// Commit the value (which allows consumers to apply their validation)
+		handleOnCommit( event );
```

Comment:
Oh, good catch about the case of `onChange` being called with the constrained value only when the input gains focus again!

Reply:
Good point, @stokesman — makes a lot of sense!

I've applied your suggestion (https://github.com/WordPress/gutenberg/pull/39186/commits/cbf0beffb1d3d6d77d7d0382db7e44b1c5d06c80) and I've also added a unit test to check the behaviour and prevent future regressions (https://github.com/WordPress/gutenberg/pull/39186/commits/80982d0cb9d69d40cae4fa67bba80283ed298e18)

Reply:
Nice tests!

Something we might want to think about with regards to `event.target.validity.valid` is that it will be `false` for `step` mismatches and [those can occur even without typing them](34566) but I don't really know if it's a concern here or not.

Reply:
That's a good point!

I also think it's not really a concern here — I'd personally keep it as-is and only make adjustments if needed.

After all, if `event.target.validity.valid` is too strict, folks can also implement their own "validation" (which is that's happening at the moment)